chainer/_link.py

```diff
@@ -0,0 +1,370 @@
+
+    @property
+    def device(self):
+        """Returns the sole device"""
+        if self._device is None:
+            visitor = _GetSoleDeviceVisitor()
+            self.visit_device_residents(visitor)
+            sole_device = visitor.sole_device
+            if sole_device is None:
+                self._device = self._MIXED_DEVICE
+            else:
+                self._device = sole_device
+
+        if self._device is self._MIXED_DEVICE:
+            return None
+        return self._device
+
+    @property
+    def xp(self):
+        # type: () -> types.Xp
+        """Array module for this link.
+
+        Depending on which of CPU/GPU this link is on, this property returns
+        :mod:`numpy` or :mod:`cupy`.
+
+        """
+        device = self.device
+        if device is None:
+            return None
+        return device.xp
+
+    @utils.final(action=DeprecationWarning)
+    def to_cpu(self):
+        # type: () -> 'DeviceResident'
+        """Copies parameter variables and persistent values to CPU.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to CPU, the link implementation must
+        override :meth:`Link.to_device` to do so.
+
+        Returns: self
+
+        """
+        visitor = _ToDeviceVisitor(
+            backend.CpuDevice(),
+            entry_method_info=('to_cpu', {}),
+            skip_between_cupy_devices=True)
+        self.__to_device(visitor)
+        return self
+
+    @utils.final(action=DeprecationWarning)
+    def to_gpu(
+            self,
+            device=None,  # type: tp.Optional[types.CudaDeviceSpec]
+    ):
+        # type: (...) -> 'DeviceResident'
+        """Copies parameter variables and persistent values to GPU.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to GPU, the link implementation must
+        override :meth:`Link.to_device` to do so.
+
+        Args:
+            device: Target device specifier. If omitted, the current device is
+                used.
+
+        Returns: self
+
+        """
+        cuda.check_cuda_available()
+        cuda_device = cuda._get_device_or_current(device)
+        device = chainer.backends.cuda.GpuDevice(cuda_device)
+        visitor = _ToDeviceVisitor(
+            device,
+            entry_method_info=('to_gpu', {'device': device}),
+            skip_between_cupy_devices=True)
+        self.__to_device(visitor)
+        return self
+
+    @utils.final(action=DeprecationWarning)
+    def to_intel64(self):
+        # type: () -> 'DeviceResident'
+        """Copies parameter variables and persistent values to CPU."""
+        intel64.check_ideep_available()
+        visitor = _ToDeviceVisitor(
+            chainer.get_device(intel64),
+            entry_method_info=('to_intel64', {}))
+        self.__to_device(visitor)
+        return self
+
+    def to_chx(self):
```

Comment:
Isn't it needed to be final? (ditto for from_chx)

Reply:
Fixed.